spring-kafka/README.md

```diff
@@ -0,0 +1,357 @@
+    </map>
+  </argument>
+</bean>
+
+<!-- Producer Kafka Template -->
+<bean id="kafkaProxyProducerKafkaTemplate" class="org.springframework.kafka.core.KafkaTemplate">
+  <argument> <!-- type converter used on this argument -->
+    <bean class="org.springframework.kafka.core.DefaultKafkaProducerFactory">
+      <argument ref="kafkaProducerProperties"/>
+      <argument> <!-- type converter used on this argument -->
+        <bean class="org.apache.kafka.common.serialization.StringSerializer" />
+      </argument>
+      <argument> <!-- type converter used on this argument -->
+        <bean class="org.hitachivantara.spring.kafka.protobuf.KafkaProtobufSerializer" />
+      </argument>
+    </bean>
+  </argument>
+  <argument type="java.lang.Boolean" value="true" /> <!-- auto flush -->
+</bean> 
+```
+
+A `BaseMessageProducer` class was added to be able to produce messages just by defining a bean for that class. That way the usage of 
+the Kafka Template in our code is black boxed. To be able to use it a bean like this must be defined:
+```xml
+<bean id="mySimpleKafkaProducerBean" class="org.hitachivantara.spring.kafka.producer.BaseMessageProducer">
+  <property name="kafkaTemplate" ref="kafkaProxyProducerKafkaTemplate"/>
+  <property name="kafkaTopic" value="myKafkaTopic"/>
+</bean>
+```
+
+### Type converter
+Notice that the `org.springframework.kafka.core.KafkaTemplate` entry in the type converter is needed in order to 
+inject the example bean `kafkaProxyProducerKafkaTemplate` in our applications beans like this:
+```xml
+  <bean id="ourApplicationBean" class="com.hitachivantara.ApplicationBean">
+    <property name="kafkaTemplate" ref="kafkaProxyProducerKafkaTemplate"/> <!-- type converter used on this argument -->
+  </bean>
+``` 
+```java
+public class ApplicationBean {
+  private KafkaTemplate<String, Message> kafkaTemplate;
+  
+  public void setKafkaTemplate( KafkaTemplate<String, Message> kafkaTemplate ) {
+    this.kafkaTemplate = kafkaTemplate;
+  }
+}
+```
+
+### KafkaTemplate
+The easiest way to send messages is using the Spring Kafka Template API defined in `org.springframework.kafka.core.KafkaTemplate`.
+That class has two constructors. Both accept as the first argument an interface `org.apache.kafka.clients.producer.Producer` implementation instance
+and the second optional argument is a `boolean` to configure the auto flush ability of the Kafka Template instance.
+
+#### Producer Factory
+Spring Kafka provides a default implementation for that interface `org.springframework.kafka.core.DefaultKafkaProducerFactory`. We
+default to use that, for which we must provide an `HashMap` with the Kafka connection and library properties.
+
+Those properties are defined in this example in the `kafkaProducerProperties`, and notice that only the basic ones are being used.
+Kafka and Spring Kafka documentation should be used to check which properties exist and how they may influence the application.
+
+#### Producer Key & Value Serializer
```

Comment:
Must we explicitly declare the serializers for keys and values? Are the serializer not shared and chosen according to the object type? 🤔

Reply:
The Listener classes are generic types which demand to know what are the key and value types. Without the serializer and deserializer there is no way for the framework to provide us the key and value object because the messages are not able to self introspect their type.

Notice however that this ability was introduced with some header fields in the message after Kafka Client 0.11 (this is version we're using... so no way to use that feature).
https://docs.spring.io/spring-kafka/reference/html/#support-for-headers
https://docs.spring.io/spring-kafka/reference/html/#headers
https://docs.spring.io/spring-kafka/reference/html/#delegating-serialization